sklearn/metrics/_dist_metrics.pyx

```diff
@@ -252,6 +256,18 @@ cdef class DistanceMetric:
         # In Minkowski special cases, return more efficient methods
         if metric is MinkowskiDistance:
             p = kwargs.pop('p', 2)
+            w = kwargs.pop('w', None)
+            if w is not None:
+                # Be consistent with scipy 1.8 conventions: in scipy 1.8,
+                # 'wminkowski' was removed in favor adding the possibility
+                # of passing a weight vector directly to 'minkowski', however
+                # the new weights apply to the absolute differences raised to
+                # the p power instead of the absolute difference as in
+                # previous versions of scipy.
+                # WMinkowskiDistance in sklearn implements the weighting
+                # scheme of the old 'wminkowski' in scipy < 1.8 and the
+                # following adaptation:
+                return WMinkowskiDistance(p, w ** (1/p), **kwargs)
```

Comment:
The issue then is that the user can create a `WMinkowskiDistance(p, q, **kwargs)` themself, which would be different than `get_metric("wminkowski", p, w)`, and kinda similar to `get_metric("minkowski", p, w)`, but with a transformation on `w`. I think this can cause quite a bit of confusion, and I'm not sure how to solve it.

If we remove `"wminkowski"`, then it'd be more consistent I guess

Reply:
I can rewrite `MinkowskiDistance` and `WMinkowskiDistance` to implement the new scheme in `MinkowskiDistance` directly and make `WMinkowskiDistance` inherit from `MinkowskiDistance` with the transformation of the weight in its constructor and leave `DistanceMetric.get_metric` to be more straightforward.

And then later we can deprecate `"wminkowski"` (but probably after a year or two, once scipy 1.8 is commonly installed).

Reply:
I think it will be easier to do this change once we no longer support versions of scipy (pre-1.6) where `wminkowski` is not deprecated.

I think this PR is good enough for now. WDYT?

Reply:
> I think this PR is good enough for now. WDYT?

I agree with you and believe the steps you gave in the first comment are proper ones to take in another PR.

Reply:
I'm looking at the metric across the board, and here's what I see:

`_dist_metrics.pyx:558-594`

``` python
# W-Minkowski Distance
#  d = sum(w_i^p * (x_i^p - y_i^p)) ^ (1/p)
cdef class WMinkowskiDistance(DistanceMetric):
    r"""Weighted Minkowski Distance

    .. math::
       D(x, y) = [\sum_i |w_i * (x_i - y_i)|^p] ^ (1/p)
```
the comment and the docstring are inconsistent, but the docstring seems to be consistent with the implementation:

``` python
        for j in range(size):
            d += pow(self.vec[j] * fabs(x1[j] - x2[j]), self.p)
```

Now this PR is making this change (on the `minkowski` part):

```
    "minkowski"     MinkowskiDistance     p, w      ``sum(w * |x - y|^p)^(1/p)``
    "wminkowski"    WMinkowskiDistance    p, w      ``sum(|w * (x - y)|^p)^(1/p)``
```

and the two are just different and not consistent with one another.

Since scipy has deprecated `wminkowski` already for a while, and now it's removed, I'd say it makes sense for this PR to deprecate `WMinkowskiDistance`, and make everything consistent with scipy?

Reply:
Actually, if one of `w` component is negative, `WMinkowskiDistance`'s call to `pow` might crash.

Reply:
> Since scipy has deprecated wminkowski already for a while, and now it's removed, I'd say it makes sense for this PR to deprecate WMinkowskiDistance, and make everything consistent with scipy? Especially since from what I see, it seems to be the case that we also have been a little bit flaky on how we incorporate w there, and now is a good time to just stick to one model.

I intended this PR to be minimal so as to be backportable to 1.0.2.

I would rather not deprecate a scikit-learn component in a scikit-learn minor release, even if that component maps more or less directly to a deprecated components in a dependency.

I can do the proposed change but I think it should target 1.1 and then 1.0.2 will have 2 broken tests under scipy 1.8 but maybe we don't care. Or maybe we care because it can make the release process automation fail...

Reply:
> the comment and the docstring are inconsistent, but the docstring seems to be consistent with the implementation

The inline comment is just plain wrong. Let me delete it, it's just useless.

> Now this PR is making this change (on the minkowski part):
>    "minkowski"     MinkowskiDistance     p, w      ``sum(w * |x - y|^p)^(1/p)``
>   "wminkowski"    WMinkowskiDistance    p, w      ``sum(|w * (x - y)|^p)^(1/p)``
> and the two are just different and not consistent with one another.

This is correct. `DistanceMetric.get_metric("minkowski", p, w)` is not the same as `DistanceMetric.get_metric("minkowski", p, w)` as you need to take the p-root or p-power of one w to get the equivalent `w` of the other. This is all consistent with what scipy does.